### PR TITLE
[EC-152] Hide Subscription/Billing information for Provider-managed organizations

### PR DESCRIPTION
**Pull request by eliykat**
## Type of change
- [ ] Bug fix
- [x] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other

## Objective

Supporting changes for https://github.com/bitwarden/web/pull/1614 - add equivalent permission checks to server.

## Code changes

* **src/Core/Context/CurrentContext.cs** - add `ManageBilling` permission check, equivalent to the jslib `organization.canManageBilling`
* **src/Api/Controllers/OrganizationsController.cs** - update the permission check on all org billing endpoints

## Testing requirements

See main ticket

## Before you submit
- [x] I have checked for formatting errors (`dotnet tool run dotnet-format --check`) (required)
- [ ] If making database changes - I have also updated Entity Framework queries and/or migrations
- [ ] I have added **unit tests** where it makes sense to do so (encouraged but not required)
- [ ] This change requires a **documentation update** (notify the documentation team)
- [ ] This change has particular **deployment requirements** (notify the DevOps team)

### CODE REVIEW

File: src/Core/Context/CurrentContext.cs [hunk 1]
```diff
@@ -261,7 +261,7 @@ public async Task<bool> OrganizationOwner(Guid orgId)
 
             if (Providers.Any())
             {
-                return (await GetProviderOrganizations()).Any(po => po.OrganizationId == orgId);
+                return (await ProviderUserForOrgAsync(orgId));
```
Comment (CarleyDiaz-Bitwarden) [inline src/Core/Context/CurrentContext.cs:264]: Nice! I like this change 👍

Comment (Hinton) [inline src/Core/Context/CurrentContext.cs:264]:
```suggestion
                return await ProviderUserForOrgAsync(orgId);
```

File: src/Core/Context/CurrentContext.cs [hunk 2]
```diff
@@ -360,6 +360,12 @@ public async Task<bool> ManageResetPassword(Guid orgId)
                         && (o.Permissions?.ManageResetPassword ?? false)) ?? false);
         }
 
+        public async Task<bool> ManageBilling(Guid orgId)
+        {
+            return await OrganizationOwner(orgId) &&
+                (await ProviderUserForOrgAsync(orgId) || ProviderIdForOrg(orgId) == null);
```
Comment (Hinton) [inline src/Core/Context/CurrentContext.cs:366]: Should we extract this to a helper method to make it more clear what we are checking? 

`ManagedByProvider(orgId)`

File: src/Core/Context/CurrentContext.cs (hunk 1 above)
Comment (eliykat) [reply src/Core/Context/CurrentContext.cs:264]: I want to blame this on linting, but I'm not sure if I can :D

File: src/Core/Context/CurrentContext.cs (hunk 2 above)
Comment (eliykat) [reply src/Core/Context/CurrentContext.cs:366]: This is the helper method, I'm not sure there's much point in adding yet more methods to `CurrentContext` to abstract this further.

This doesn't feel significantly better to me:
```
return await OrganizationOwner(orgId) &&
                (await ProviderUserForOrgAsync(orgId) || !(await ManagedByProvider(orgId));
```

and if you try to encapsulate more than that, you need to name it something like

```
return await OrganizationOwner(orgId) &&
                !(await ManagedByProviderAndImNotAProviderUser(orgId));
```

File: src/Api/Controllers/OrganizationsController.cs [hunk 3]
```diff
@@ -230,21 +230,22 @@ public async Task<OrganizationResponseModel> PostLicense(OrganizationCreateLicen
...
 
-            var updatebilling = !_globalSettings.SelfHosted && (model.BusinessName != organization.BusinessName ||
+            var updateBilling = !_globalSettings.SelfHosted && (model.BusinessName != organization.BusinessName ||
                 model.BillingEmail != organization.BillingEmail);
 
-            await _organizationService.UpdateAsync(model.ToOrganization(organization, _globalSettings), updatebilling);
+            if ((updateBilling && !await _currentContext.ManageBilling(orgIdGuid)) || !await _currentContext.OrganizationOwner(orgIdGuid))
+            {
+                throw new NotFoundException();
+            }
```
[the remark this answers is not in the document]
Comment (eliykat) [reply src/Api/Controllers/OrganizationsController.cs]: Have tried to break this down a little more.

**Review by Hinton — DISMISSED** [review]

**Comment by eliykat** [comment]
@Hinton thought of how to make `ManageBilling` a bit clearer, per your feedback. Ternary if > boolean logic

**Comment by eliykat** [comment]
Holding as explained [here](https://github.com/bitwarden/web/pull/1614#issuecomment-1120650193), if this is merged without web then it'll break things.

File: src/Core/Context/CurrentContext.cs [hunk 4]
```diff
@@ -362,8 +362,9 @@ public async Task<bool> ManageResetPassword(Guid orgId)
 
         public async Task<bool> ManageBilling(Guid orgId)
         {
-            return await OrganizationOwner(orgId) &&
-                (await ProviderUserForOrgAsync(orgId) || ProviderIdForOrg(orgId) == null);
+            return ProviderIdForOrg(orgId) == null
```
Comment (Hinton) [inline src/Core/Context/CurrentContext.cs:365]: I feel like it's still a bit non-intuitive what ProviderIdForOrg does, how about extracting it to a variable?
```suggestion
			var managedByProvider = ProviderIdForOrg(orgId) == null;
            return managedByProvider
```

Comment (eliykat) [reply src/Core/Context/CurrentContext.cs]: Done, but flipped it for readability

**Review by Hinton — APPROVED** [review]

**Pull request merged**